File: frontend/src/api/portfolio.ts
import { api, apiUrl } from "./client";
import { getAccessToken } from "@/lib/tokens";

export interface OptimizeRequest {
  tickers: string[];
  use_db_data?: boolean;
  lookback_days?: number;
  as_of?: string;
  expected_returns?: number[];
  cov_daily?: number[][];
  daily_returns?: number[][];
  risk_free_rate?: number;
  method?: "slsqp" | "qp";
  min_stock_sd?: number;
  apply_min_sd_constraint?: boolean;
  apply_return_floor?: boolean;
  allow_shorting?: boolean;
}

export interface OptimizeResponse {
  success: boolean;
  method: string;
  message: string;
  weights: Record<string, number>;
  sharpe: number;
  expected_return: number;
  volatility: number;
  sum_weights: number;
  risk_free_rate: number;
  min_stock_sd: number | null;
  tickers: string[];
  capm_expected_return: Record<string, number>;
  annual_volatility: Record<string, number>;
  beta: Record<string, number>;
  cov_daily: number[][] | null;
  run_id: number | null;
}

export interface FrontierPoint {
  target_return: number;
  volatility: number;
  weights: number[];
}

export interface FrontierResponse {
  tickers: string[];
  points: FrontierPoint[];
  tangency_return: number;
  tangency_volatility: number;
  tangency_weights: Record<string, number>;
}

export interface MetricsRequest {
  tickers: string[];
  weights: number[];
  risk_free_rate?: number;
  use_db_data?: boolean;
  expected_returns?: number[];
  cov_daily?: number[][];
  daily_returns?: number[][];
}

export interface MetricsResponse {
  sharpe: number;
  expected_return: number;
  volatility: number;
  sum_weights: number;
  risk_contribution: Record<string, number>;
  var_95_daily: number;
  var_95_10d: number;
  cvar_95_daily: number;
}

export interface PortfolioRun {
  id: number;
  run_at: string;
  method: string;
  risk_free_rate: number;
  expected_return: number | null;
  volatility: number | null;
  sharpe: number | null;
  var_95: number | null;
  success: boolean;
  weights: Record<string, number> | null;
}

// Saved-portfolio CRUD types (Loay slide 1)
export interface PortfolioHolding {
  ticker: string;
  weight: number;
  purchase_date: string | null;
  purchase_price: number | null;
}

export interface AddHoldingPayload {
  ticker: string;
  purchase_date?: string;
  purchase_price?: number;
}

export interface SavedPortfolio {
  id: number;
  name: string;
  description: string | null;
  initial_capital: number | null;
  target_loss_threshold: number | null;
  holdings: PortfolioHolding[];
  status: "active" | "inactive";
  holding_count: number;
  total_weight: number;
  needs_recompute: boolean | null;
  created_at: string;
  updated_at: string;
}

export interface CreatePortfolioPayload {
  name: string;
  initial_capital: number | null;
}

export interface UpdatePortfolioPayload {
  name?: string;
  description?: string;
  initial_capital?: number;
  target_loss_threshold?: number;
}

export interface ComputeWeightsPayload {
  method?: "slsqp" | "qp";
  risk_free_rate?: number;
  lookback_days?: number;
  apply_min_sd_constraint?: boolean;
  apply_return_floor?: boolean;
  allow_shorting?: boolean;
}

export interface ComputeWeightsResult {
  portfolio_id: number;
  run_id: number;
  success: boolean;
  method: string;
  message: string;
  sharpe: number;
  expected_return: number;
  volatility: number;
  risk_free_rate: number;
  weights: Record<string, number>;
  annual_volatility: Record<string, number>;
  beta: Record<string, number>;
  capm_expected_return: Record<string, number>;
}

export const PortfolioAPI = {
  optimize: (req: OptimizeRequest) =>
    api<OptimizeResponse>("/portfolio/optimize", { method: "POST", body: req }),

  frontier: (req: OptimizeRequest, nPoints = 50) =>
    api<FrontierResponse>(`/portfolio/frontier?n_points=${nPoints}`, {
      method: "POST",
      body: req,
    }),

  metrics: (req: MetricsRequest) =>
    api<MetricsResponse>("/portfolio/metrics", { method: "POST", body: req }),

  runs: (limit = 20) => api<PortfolioRun[]>(`/portfolio/runs?limit=${limit}`),

  // Saved-portfolio CRUD (Loay slide 1)
  listSaved: () => api<SavedPortfolio[]>("/portfolio/"),

  create: (payload: CreatePortfolioPayload) =>
    api<SavedPortfolio>("/portfolio/", {
      method: "POST",
      body: { ...payload, holdings: [] },
    }),

  update: (id: number, patch: UpdatePortfolioPayload) =>
    api<SavedPortfolio>(`/portfolio/${id}`, { method: "PATCH", body: patch }),

  remove: (id: number) =>
    api<void>(`/portfolio/${id}`, { method: "DELETE" }),

  // Holdings (per-stock) + compute weights (closes the "active" loop)
  addHolding: (id: number, payload: AddHoldingPayload) =>
    api<SavedPortfolio>(`/portfolio/${id}/holdings`, {
      method: "POST",
      body: payload,
    }),

  removeHolding: (id: number, ticker: string) =>
    api<SavedPortfolio>(`/portfolio/${id}/holdings/${encodeURIComponent(ticker)}`, {
      method: "DELETE",
    }),

  computeWeights: (id: number, payload: ComputeWeightsPayload = {}) =>
    api<ComputeWeightsResult>(`/portfolio/${id}/compute`, {
      method: "POST",
      body: payload,
    }),

  getOne: (id: number) =>
    api<SavedPortfolio>(`/portfolio/${id}`),

  reportPdfUrl: (runId: number, locale: "ar" | "en") =>
    // GET endpoint: we expose a URL plus a token the browser can attach via Authorization.
    // Since <a href> can't carry headers, callers use downloadReport() which fetches
    // the blob and triggers a local download.
    apiUrl(`/portfolio/runs/${runId}/report.pdf?locale=${locale}`),

  async downloadReport(runId: number, locale: "ar" | "en"): Promise<void> {
    const blob = await api<Blob>(`/portfolio/runs/${runId}/report.pdf?locale=${locale}`, {
      blob: true,
    });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `portfolio-run-${runId}-${locale}.pdf`;
    document.body.appendChild(a);
    a.click();
    a.remove();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
  },
};

/** For code-splitting-free import of the access token getter (used by <img> etc.). */
export const currentAccessToken = (): string | null => getAccessToken();

File: backend/app/schemas/portfolio.py
"""Pydantic schemas for the portfolio optimization API."""
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


# ---------------------------------------------------------------------------
# Shared inputs — either DB-backed (just tickers) or manual (full matrices)
# ---------------------------------------------------------------------------
class OptimizeRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=2, max_length=100)

    # Option A: DB-backed — leave returns/cov/rf empty and set use_db_data
    use_db_data: bool = Field(
        True,
        description=(
            "If true, the server fetches prices + computes CAPM/covariance from "
            "the DB. If false, you must provide daily_returns or cov_daily below."
        ),
    )
    lookback_days: int | None = Field(
        None, ge=30, le=10 * 252, description="Override admin_config.lookback_days."
    )
    as_of: date | None = Field(
        None,
        description="Treat this date as 'today' for analytics (for back-testing).",
    )

    # Option B: manual inputs (legacy parity with root-level API)
    expected_returns: list[float] | None = None
    cov_daily: list[list[float]] | None = None
    daily_returns: list[list[float]] | None = None
    risk_free_rate: float | None = Field(
        None, ge=0, le=0.5,
        description="Override admin_config.risk_free_rate.",
    )

    # Solver / constraint controls
    method: Literal["slsqp", "qp"] = "qp"
    min_stock_sd: float | None = Field(
        None,
        description="Constraint cap σ_p ≤ this. Default = MIN of per-stock annualized SDs.",
    )
    apply_min_sd_constraint: bool = True
    apply_return_floor: bool = True
    allow_shorting: bool = False

    @field_validator("tickers")
    @classmethod
    def _unique(cls, v: list[str]) -> list[str]:
        seen: set[str] = set()
        out = []
        for t in v:
            t2 = t.strip()
            if t2 in seen:
                continue
            seen.add(t2)
            out.append(t2)
        return out


class OptimizeResponse(BaseModel):
    success: bool
    method: str
    message: str
    weights: dict[str, float]
    sharpe: float
    expected_return: float
    volatility: float
    sum_weights: float

    # Echo back the inputs the server used so the frontend can display them
    risk_free_rate: float
    min_stock_sd: float | None
    tickers: list[str]
    capm_expected_return: dict[str, float]
    annual_volatility: dict[str, float]
    beta: dict[str, float]
    cov_daily: list[list[float]] | None = None
    run_id: int | None = None


class FrontierPoint(BaseModel):
    target_return: float
    volatility: float
    weights: list[float]


class FrontierResponse(BaseModel):
    tickers: list[str]
    points: list[FrontierPoint]
    tangency_return: float
    tangency_volatility: float
    tangency_weights: dict[str, float]


class MetricsRequest(BaseModel):
    """For the live slider — recompute on a user-supplied weight vector."""
    tickers: list[str]
    weights: list[float]
    expected_returns: list[float] | None = None
    cov_daily: list[list[float]] | None = None
    daily_returns: list[list[float]] | None = None
    risk_free_rate: float | None = None
    use_db_data: bool = True


class MetricsResponse(BaseModel):
    sharpe: float
    expected_return: float
    volatility: float
    sum_weights: float
    risk_contribution: dict[str, float]
    var_95_daily: float
    var_95_10d: float
    cvar_95_daily: float


# ---------------------------------------------------------------------------
# Saved-portfolio CRUD
# ---------------------------------------------------------------------------
class HoldingIn(BaseModel):
    ticker: str
    weight: float = Field(..., ge=0, le=1)


class SavePortfolio(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    initial_capital: float | None = Field(None, ge=0)
    target_loss_threshold: float | None = Field(None, ge=0, le=1)
    # holdings is optional: Loay's "create new portfolio" modal submits just
    # name + amount. The optimize flow still submits a full holdings list with
    # weights summing to 1 — both cases handled by the /portfolio/ POST.
    holdings: list[HoldingIn] = Field(default_factory=list, max_length=269)


class UpdatePortfolioRequest(BaseModel):
    """PATCH body: every field optional; omitted fields are left untouched."""
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    initial_capital: float | None = Field(None, ge=0)
    target_loss_threshold: float | None = Field(None, ge=0, le=1)


class AddHoldingRequest(BaseModel):
    """Add one stock to a portfolio's holdings (no weight — set via compute)."""
    ticker: str = Field(..., min_length=1, max_length=12)


class ComputeWeightsRequest(BaseModel):
    """Trigger Markowitz optimization on a portfolio's existing holdings."""
    method: Literal["slsqp", "qp"] = "qp"
    risk_free_rate: float | None = Field(None, ge=0, le=0.5)
    lookback_days: int | None = Field(None, ge=30, le=10 * 252)
    apply_min_sd_constraint: bool = True
    apply_return_floor: bool = True
    allow_shorting: bool = False


class ComputeWeightsResponse(BaseModel):
    """Return from POST /portfolio/{id}/compute — mirrors OptimizeResponse but
    tied to the portfolio (weights have been persisted into its holdings)."""
    portfolio_id: int
    run_id: int
    success: bool
    method: str
    message: str
    sharpe: float
    expected_return: float
    volatility: float
    risk_free_rate: float
    weights: dict[str, float]
    annual_volatility: dict[str, float]
    beta: dict[str, float]
    capm_expected_return: dict[str, float]


class PortfolioOut(BaseModel):
    id: int
    name: str
    description: str | None
    initial_capital: float | None
    target_loss_threshold: float | None
    holdings: list[HoldingIn]
    # Derived status per PPTX slide 1 ("Active" = picked stocks AND computed weights)
    status: Literal["active", "inactive"]
    holding_count: int
    total_weight: float
    # Set only on PATCH responses when initial_capital changed on a portfolio
    # that already has holdings — the frontend surfaces a "recompute weights"
    # banner in that case.
    needs_recompute: bool | None = None
    created_at: datetime
    updated_at: datetime


class PortfolioRunOut(BaseModel):
    id: int
    run_at: datetime
    method: str
    risk_free_rate: float
    expected_return: float | None
    volatility: float | None
    sharpe: float | None
    var_95: float | None
    success: bool
    weights: dict[str, float] | None


# ---------------------------------------------------------------------------
# Phase 2: VaR Full (all methods + FRTB ES)
# ---------------------------------------------------------------------------
class VarFullRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=2)
    weights: list[float]
    confidence: float = Field(0.95, ge=0.9, le=0.999)
    horizon_days: int = Field(1, ge=1, le=250)
    n_paths: int = Field(10_000, ge=1000, le=100_000)
    include_garch: bool = Field(True, description="Include GARCH-t VaR (slower)")
    include_frtb: bool = Field(True, description="Include Basel FRTB ES at 97.5%")

    # Data source
    use_db_data: bool = True
    daily_returns: list[list[float]] | None = None
    cov_daily: list[list[float]] | None = None
    lookback_days: int | None = None
    risk_free_rate: float | None = None


class VarMethodResult(BaseModel):
    method: str
    confidence: float
    horizon_days: int
    var_loss: float
    cvar_loss: float | None
    portfolio_mean_daily: float
    portfolio_vol_daily: float
    simulated_paths: int | None = None


class FrtbResult(BaseModel):
    es_975: float
    var_975: float
    horizon_days: int
    confidence: float
    garch_params: dict | None = None
    n_paths: int
    fallback: str | None = None


class VarFullResponse(BaseModel):
    tickers: list[str]
    methods: dict[str, VarMethodResult]
    frtb: FrtbResult | None = None


# ---------------------------------------------------------------------------
# Phase 2: VaR Backtest (Kupiec + Christoffersen)
# ---------------------------------------------------------------------------
class BacktestRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=2)
    weights: list[float]
    confidence: float = Field(0.95, ge=0.9, le=0.999)
    var_method: Literal["historical", "parametric", "monte_carlo"] = "parametric"
    window: int = Field(252, ge=60, le=2520, description="Rolling lookback window")
    significance: float = Field(0.05, ge=0.01, le=0.10)

    use_db_data: bool = True
    daily_returns: list[list[float]] | None = None
    lookback_days: int | None = None
    risk_free_rate: float | None = None


class BacktestResponse(BaseModel):
    var_method: str
    confidence: float
    n_observations: int
    n_violations: int
    expected_violations: float
    violation_rate: float
    expected_violation_rate: float

    kupiec_lr: float
    kupiec_pval: float
    pass_kupiec: bool

    christoffersen_lr: float | None
    christoffersen_pval: float | None
    pass_christoffersen: bool | None

    independence_lr: float | None
    independence_pval: float | None


# ---------------------------------------------------------------------------
# Phase 2: Correlation + Covariance Quality
# ---------------------------------------------------------------------------
class CorrelationRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=2)
    use_db_data: bool = True
    daily_returns: list[list[float]] | None = None
    lookback_days: int | None = None
    risk_free_rate: float | None = None


class CovQualityReport(BaseModel):
    condition_number: float
    is_psd: bool
    min_eigenvalue: float
    max_eigenvalue: float
    trace: float
    determinant_log: float
    suggested_shrinkage: float


class CorrelationResponse(BaseModel):
    tickers: list[str]
    correlation: list[list[float]]
    correlation_parity_pass: bool
    correlation_max_diff: float
    cov_daily: list[list[float]]
    cov_daily_shrunk: list[list[float]]
    shrinkage_intensity: float
    cov_quality: CovQualityReport
    beta_raw: dict[str, float]
    beta_blume: dict[str, float]
